modules/dnn/src/tensorflow/tf_io.cpp

```diff
@@ -52,6 +52,18 @@ bool ReadProtoFromBinaryFileTF(const char* filename, Message* proto) {
     return success;
 }
 
+bool ReadProtoFromBinaryBufferTF(const char* data, size_t len, Message* proto) {
```

Comment:
I think it's better to take existing two methods, remove `std::ifstream` creation from them and replace `const char* filename` argument to `std::istream& is` one.
In result, we'll have something like 

```cpp
bool ReadProtoFromTextFileTF(std::istream& s, Message* proto) {
    IstreamInputStream input(&is);
    return google::protobuf::TextFormat::Parse(&input, proto);
}
```
```cpp
bool ReadProtoFromTextFileTF(const char* buf, size_t len, Message* proto) {
    std::istringstream s(std::string(buf, len));
    return ReadProtoFromTextFileTF(s, proto);
}
```
```cpp
bool ReadProtoFromTextFileTF(const char* filename, Message* proto) {
    std::ifstream fs(filename, std::ios::in);
    CHECK(fs.is_open()) << "Can't open \"" << filename << "\"";
    return ReadProtoFromTextFileTF(fs, proto);
}
```

Reply:
The problem with std::istringstream s(std::string(buf, len)) is that, in some cases, it creates a copy of the buffer into a std::string.

However ArrayInputStream input(data, len) guarantees that there is no copy
https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.io.zero_copy_stream_impl_lite#ArrayInputStream

Reply:
@r2d3, we just vote to reduce a code duplication. There are two methods `ReadProtoFromBinaryFileTF` and `ReadProtoFromBinaryBufferTF` are different only at the beginning for now.

Reply:
OK I will reduce code duplication

Reply:
I put the common code in `ReadProtoFromBinaryTF`.

I simplified the code in other functions. `raw_input` and `coded_input` are now created on stack to avoid new/delete.